### PR TITLE
[BEAM-1867] Use step-derived PCollection names in Dataflow

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

R: @bjchambers 

This mitigates an issue in Dataflow. I also removed some checked exceptions that are never caught and probably never should be.

I have empirically checked that the element counts and byte sizes are restored by this change, and added unit tests to the translator. Integration tests TBD.

### CODE REVIEW

**Comment by kennknowles** [comment]
Run Dataflow ValidatesRunner

**Comment by kennknowles** [comment]
retest this please

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java [hunk 1]
```diff
@@ -656,7 +656,12 @@ private long addOutput(PValue value, Coder<?> valueCoder) {
 
       Map<String, Object> outputInfo = new HashMap<>();
       addString(outputInfo, PropertyNames.OUTPUT_NAME, Long.toString(id));
-      addString(outputInfo, PropertyNames.USER_NAME, value.getName());
+
+      String stepName = getString(properties, PropertyNames.USER_NAME);
+      String generatedName = String.format(
+          "%s.out%d", stepName, outputInfoList.size());
```
Comment (tgroh) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java]: Tuples are arbitrarily ordered; do we care that we don't have a strong definition of what the expected names are? What about named TupleTags (with non-generated IDs)?

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java [hunk 2]
```diff
@@ -772,6 +772,100 @@ public void testInaccessibleProvider() throws Exception {
...
+            Structs.getListOfMaps(
+                step.getProperties(),
+                PropertyNames.OUTPUT_INFO,
+                null).get(0),
+        PropertyNames.USER_NAME);
+
+    // This is the calculated name that must actually be used
+    String calculatedName = getString(step.getProperties(), PropertyNames.USER_NAME) + ".out0";
+
+    assertThat(userSpecifiedName, equalTo(calculatedName));
```
Comment (tgroh) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java]: What about something you've named with `PCollection.setName`

Comment (kennknowles) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java]: This one is named "foobizzle". Do you mean something that I didn't name? (and maybe that comment applies more to the other test?)

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java (hunk 1 above)
Comment (kennknowles) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java]: The tags themselves also have to be ignored - the unit tests cover both TupleTags with specified strings and TupleTags without specified strings.

**Pull request merged**